File: converters/video.py
"""
video.py - converts video to gif using ffmpeg
"""

import os
import subprocess

from rich.console import Console
from rich.prompt import Prompt

from ui.theme import Theme

console = Console()


def check_ffmpeg():
    # verify ffmpeg is installed
    try:
        subprocess.run(
            ["ffmpeg", "-version"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            check=True,
        )
        return True
    except (FileNotFoundError, subprocess.CalledProcessError):
        return False


def conv_video(path):
    console.rule(f"[bold {Theme.HEADER}]VIDEO TO GIF[/bold {Theme.HEADER}]")

    # check dependencies
    with console.status(f"[{Theme.INFO}]Checking FFmpeg...[/{Theme.INFO}]"):
        if not check_ffmpeg():
            console.print(
                f"[bold {Theme.ERROR}]ERROR:[/bold {Theme.ERROR}] FFmpeg is not installed or not in PATH."
            )
            return
        console.print(f"[{Theme.SUCCESS}]✓ FFmpeg detected.[/{Theme.SUCCESS}]")

    # get conversion settings
    fps = Prompt.ask(
        f"[{Theme.HEADER}]Select FPS[/{Theme.HEADER}]",
        choices=["10", "24", "30"],
        default="10",
    )
    width = Prompt.ask(
        f"[{Theme.HEADER}]Select Width[/{Theme.HEADER}]",
        choices=["320", "480", "720", "1080"],
        default="480",
    )

    convert_to_gif(path, fps, width)


def convert_to_gif(path, fps, width):
    if not os.path.isfile(path):
        console.print(f"[{Theme.ERROR}]File not found[/{Theme.ERROR}]")
        return None

    # define paths
    out_dir = os.path.dirname(path) or "."
    name = os.path.splitext(os.path.basename(path))[0]
    output_gif = os.path.join(out_dir, name + ".gif")
    palette_png = os.path.join(out_dir, f"{name}_palette.png")
    filters = f"fps={fps},scale={width}:-1:flags=lanczos"

    run_kwargs = {
        "check": True,
        "stdout": subprocess.DEVNULL,
        "stderr": subprocess.STDOUT,
    }

    try:
        # process with spinner
        with console.status(
            f"[bold {Theme.INFO}]Processing (this might take a moment)...[/bold {Theme.INFO}]"
        ):
            # generate palette
            subprocess.run(
                [
                    "ffmpeg",
                    "-y",
                    "-i",
                    path,
                    "-vf",
                    f"{filters},palettegen",
                    palette_png,
                ],
                **run_kwargs,
            )
            # generate final gif
            subprocess.run(
                [
                    "ffmpeg",
                    "-y",
                    "-i",
                    path,
                    "-i",
                    palette_png,
                    "-lavfi",
                    f"{filters} [x]; [x][1:v] paletteuse",
                    output_gif,
                ],
                **run_kwargs,
            )

        console.print(
            f"\n[bold {Theme.SUCCESS}]Success![/bold {Theme.SUCCESS}] GIF saved at: {output_gif}"
        )

    except subprocess.CalledProcessError:
        console.print(
            f"[bold {Theme.ERROR}]Error: Conversion failed.[/bold {Theme.ERROR}]"
        )
        return None
    finally:
        # cleanup palette
        if os.path.exists(palette_png):
            os.remove(palette_png)

File: src/converters/video.rs
/// video.rs - converts video to gif using ffmpeg

use std::path::Path;
use std::process::Command;
use std::ffi::OsStr;

use console::style;
use dialoguer::Select;

use crate::ui::theme::Theme;

// --- check if ffmpeg is installed ---

fn check_ffmpeg() -> bool {
    Command::new("ffmpeg")
        .arg("-version")
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status()
        .map(|s| s.success())
        .unwrap_or(false)
}

// --- entry point: validate, ask settings, convert ---

pub fn conv_video(path: &str, args: Option<&[String]>) {
    println!();
    println!("{}", style("VIDEO TO GIF").fg(Theme::HEADER).bold());

    // check dependencies
    println!("{}", style("Checking FFmpeg...").fg(Theme::INFO));
    if !check_ffmpeg() {
        println!(
            "{} FFmpeg is not installed or not in PATH.",
            style("ERROR:").fg(Theme::ERROR).bold(),
        );
        return;
    }
    println!("{}", style("✓ FFmpeg detected.").fg(Theme::SUCCESS));

    // check supported format
    let ext = Path::new(path)
        .extension()
        .unwrap_or(OsStr::new(""))
        .to_string_lossy()
        .to_lowercase();

    let supported = ["mp4", "avi", "mov", "mkv", "webm", "flv"];
    if !supported.contains(&ext.as_str()) {
        println!(
            "{}",
            style(format!("Unsupported video format: .{}", ext)).fg(Theme::ERROR),
        );
        return;
    }

    // get conversion settings
    let (fps, width) = if let Some(args) = args {
        let f = args.get(0).map(|s| s.as_str()).unwrap_or("10");
        let w = args.get(1).map(|s| s.as_str()).unwrap_or("480");
        (f, w)
    } else {
        let fps_options = &["10", "24", "30"];
        let fps_idx = Select::new()
            .with_prompt(format!("{}", style("Select FPS").fg(Theme::HEADER)))
            .items(fps_options)
            .default(0)
            .interact()
            .unwrap_or(0);
        let f = fps_options[fps_idx];

        let width_options = &["320", "480", "720", "1080"];
        let width_idx = Select::new()
            .with_prompt(format!("{}", style("Select Width").fg(Theme::HEADER)))
            .items(width_options)
            .default(1)
            .interact()
            .unwrap_or(1);
        let w = width_options[width_idx];
        
        (f, w)
    };

    convert_to_gif(path, fps, width);
}

// --- two-pass ffmpeg conversion: palette → gif ---

fn convert_to_gif(path: &str, fps: &str, width: &str) {
    if !Path::new(path).is_file() {
        println!("{}", style("File not found").fg(Theme::ERROR));
        return;
    }

    // define paths
    let parent = Path::new(path).parent().unwrap_or(Path::new("."));
    let name = Path::new(path)
        .file_stem()
        .unwrap_or(OsStr::new("output"))
        .to_string_lossy();

    let output_gif = parent.join(format!("{}.gif", name));
    let palette_png = parent.join(format!("{}_palette.png", name));

    let filters = format!("fps={},scale={}:-1:flags=lanczos", fps, width);

    println!(
        "{}",
        style("Processing (this might take a moment)...").fg(Theme::INFO).bold(),
    );

    // generate palette
    let palette_result = Command::new("ffmpeg")
        .args([
            "-y", "-i", path,
            "-vf", &format!("{},palettegen", filters),
            palette_png.to_str().unwrap_or("palette.png"),
        ])
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status();

    if palette_result.map(|s| s.success()).unwrap_or(false) == false {
        println!(
            "{}",
            style("Error: Palette generation failed.").fg(Theme::ERROR).bold(),
        );
        cleanup_file(&palette_png);
        return;
    }

    // generate final gif
    let gif_result = Command::new("ffmpeg")
        .args([
            "-y", "-i", path,
            "-i", palette_png.to_str().unwrap_or("palette.png"),
            "-lavfi", &format!("{} [x]; [x][1:v] paletteuse", filters),
            output_gif.to_str().unwrap_or("output.gif"),
        ])
        .stdout(std::process::Stdio::null())
        .stderr(std::process::Stdio::null())
        .status();

    // cleanup palette
    cleanup_file(&palette_png);

    match gif_result {
        Ok(status) if status.success() => {
            let size_kb = std::fs::metadata(&output_gif)
                .map(|m| m.len() as f64 / 1024.0)
                .unwrap_or(0.0);
            println!();
            println!(
                "{} GIF saved at: {} ({:.1} KB)",
                style("Success!").fg(Theme::SUCCESS).bold(),
                output_gif.display(),
                size_kb,
            );
        }
        _ => {
            println!(
                "{}",
                style("Error: Conversion failed.").fg(Theme::ERROR).bold(),
            );
        }
    }
}

// --- helper: remove temp file if it exists ---

fn cleanup_file(path: &Path) {
    if path.exists() {
        let _ = std::fs::remove_file(path);
    }
}
